numba/core/interpreter.py

```diff
@@ -80,6 +78,229 @@ def get_assignment_source(self, destname):
         return None
 
 
+def peep_hole_list_to_tuple(func_ir):
+    """
+    This peephole rewrites a bytecode sequence new to Python 3.9 that looks
+    like e.g.:
+
+    def foo(a):
+        return (*a,)
+
+    41          0 BUILD_LIST               0
+                2 LOAD_FAST                0 (a)
+                4 LIST_EXTEND              1
+                6 LIST_TO_TUPLE
+                8 RETURN_VAL
+
+    essentially, the unpacking of tuples is written as a list which is appended
+    to/extended and then "magicked" into a tuple by the new LIST_TO_TUPLE
+    opcode.
+
+    This peephole repeatedly analyses the bytecode in a block looking for a
+    window between a `LIST_TO_TUPLE` and `BUILD_LIST` and...
+
+    1. Turns the BUILD_LIST into a BUILD_TUPLE
+    2. Sets an accumulator's initial value as the target of the BUILD_TUPLE
+    3. Searches for 'extend' on the original list and turns these into binary
+       additions on the accumulator.
+    4. Searches for 'append' on the original list and turns these into a
+       `BUILD_TUPLE` which is then appended via binary addition to the
+       accumulator.
+    5. Assigns the accumulator to the variable that exits the peephole and the
+       rest of the block/code refers to as the result of the unpack operation.
+    6. Patches up
+    """
+    _DEBUG = False
+
+    # For all blocks
+    for offset, blk in func_ir.blocks.items():
+        target_list = None
+        # keep doing the peephole rewrite until nothing is left that matches
+        while True:
+            found = False
+            region = None
+            # first try and find a matching region
+            # i.e. BUILD_LIST...<stuff>...LIST_TO_TUPLE
+            for idx, stmt in enumerate(reversed(blk.body)):
```

Comment:
Kind of a brainteaser that idx is the negative index and later it needs to be `-region[1][0] - 1`. It'll be clearer as positive indices. My suggestion: https://github.com/sklam/numba/commit/492222d0e324e8bacd0c3fa1e08263b38db2eaae, (I also moved the code into its a subroutine in the patch)

Reply:
:) this is because I originally wrote the algorithm to work against the incoming stack, so everything is/was backwards from the last emitted statement which would have been a `LIST_TO_TUPLE`. Agree with changes to positive indices, will check and, assuming fine, apply the suggested.

Reply:
https://github.com/sklam/numba/commit/492222d0e324e8bacd0c3fa1e08263b38db2eaae is applied to this branch.